Iterate over different seeds

brendanreardon opened the pull request:
Updating this implementation of FACETS to include iterating across multiple seeds to observe the stability of called solutions for a given tumor-normal pair. The following specific changes were made:

**Iterating over many seeds**
Users can provide an initial (numeric) seed and how many iterations that they would like to perform. Subsequent integers based on the number of requested iterations are used as additional seeds. Results from these runs are reported and plotted in the `.facets_iterations.txt` and `.facets_iterations.pdf` file, respectively. 

The number of seeds that resulted in NA purity calls are reported in the `number_iterations_with_na_purity.txt` output. For seeds that resulted in non-null purity values, the iteration with a purity closest to the median is used and reported in `seed_used.txt`. If all seeds result in null purity values, the initial seed is used. 

**Secondary changes**
- The README was updated to provide more detailed instruction for each task of this implementation, as well as references which may be useful. 
- The minimum_normal_depth parameter is passed to both the pileup and preProcSample for consistency.
- The Dockerfile was refactored some. GCC is now installed through apt-get instead of through conda due to building issues. Also the Github repository is no longer being downloaded into the docker, as it was already being installed through R devtools.

mengxiao commented on facets.wdl:
```diff
@@ -5,8 +5,8 @@ workflow facets_workflow {
 
-    File? pileup_snp_vcf = "gs://fc-f36b3dc8-85f7-4d7f-bc99-a4610229d66a/broadinstitute/reference/hg19/db/dbsnp/dbsnp_134_b37.leftAligned.vcf"
-    File? pileup_snp_vcf_index = "gs://fc-f36b3dc8-85f7-4d7f-bc99-a4610229d66a/broadinstitute/reference/hg19/db/dbsnp/dbsnp_134_b37.leftAligned.vcf.idx"
+    File? pileup_snp_vcf = "gs://fc-f36b3dc8-85f7-4d7f-bc99-a4610229d66a/common_sites/00-common_all.vcf.gz"
```
I think an inline comment is warranted explaining what database of variants this is

brendanreardon replied:
Added: https://github.com/vanallenlab/facets/commit/fe88a4afcee38f1faba55a16f74c8fe3b3df4748

mengxiao replied:
Looks like you removed the preset defaults. Are we still keeping around a mirror of the dbSNP sites somewhere in GCS that we can provide in a default config or something?

brendanreardon replied:
Yup, that will be in a default config on firecloud for our group

mengxiao commented on facets.wdl:
```diff
@@ -16,15 +16,19 @@ workflow facets_workflow {
+    Int? seed_iterations = 10
 
     Int? pileup_memoryGB = 3
     Int? pileup_diskGB = 200
```
Can this be dynamically sized?
`pileup_diskGB = ceil(1.1 * (size(normal_bam + 'G') + size(tumor_bam, 'G'))) + 2`

mengxiao commented on facets.wdl:
```diff
@@ -16,15 +16,19 @@ workflow facets_workflow {
-	Int? facets_memoryGB = 3
-	Int? facets_diskGB = 200
+    Int? facets_memoryGB = 3
+    Int? facets_diskGB = 200
```
I think this should be removed in favor of dynamic sizing

mengxiao commented on facets.wdl:
```diff
@@ -124,33 +140,42 @@ task Pileup {
+    Int seed_iterations
 
     Int? memoryGB
     Int? diskGB
```
`Int diskGB = ceil(size(pileup, 'G') * 1.1) + 2`
I actually don't know what the disk requirements of FACETS are, so some of those constants might be off

mengxiao commented on facets.wdl:
```diff
@@ -16,15 +16,19 @@ workflow facets_workflow {
 
     Int? inferwgd_memoryGB = 3
-	Int? inferwgd_diskGB = 50
+    Int? inferwgd_diskGB = 50
```
Again, use dynamic sizing

mengxiao commented on facets.wdl:
```diff
@@ -159,6 +184,7 @@ task InferWGD {
 
     Int? memoryGB
     Int? diskGB
```
`Int diskGB = ceil(size(cncf, 'G')) + 1`

mengxiao commented on README.md:
```diff
@@ -1,13 +1,85 @@
+[Relevant codeblock in WDL](https://github.com/vanallenlab/facets/blob/lab_harmonize/facets.wdl#L140-L180)
+
+### Infer Whole-genome doubling
+Infers whole genome doubling based on [Bielski CM, Zehir A, Penson AV, et al. Genome doubling shapes the evolution and prognosis of advanced cancers](https://doi.org/10.1038/s41588-018-0165-1). Calculates major copy number (MCN) estimate based on total copy number (TCN) estimate and minor copy number (LCN) estimate from FACETS and calls whole-genome doubling if the average MCN across the genome i ...
```
@brendanreardon @JakeConway I definitely don't understand this issue, but should we not exclude LCA = NA sites? Is there danger that the true LCN in these cases is > 1 (and so MCN might be overestimated)?

brendanreardon replied:
I don't think that we should be excluding LCN = NA sites because we can infer enough based on TCN alone for the scope of this task, plus we need to normalize all MCN values by their segment lengths to get the fraction of MCN > 2 across the genome. I think that the only blind spot is when the true TCN = 4 and LCN = 2, but that should indistinguishable from a diploid state. Anyway, I'll e-mail Craig to double check none the less.

mengxiao commented on facets.wdl:
```diff
@@ -41,38 +45,48 @@ workflow facets_workflow {
     output {
         File facets_pileup = Pileup.pileup
         File facets_genome_segments = FACETS.genome_segments
         File facets_diagnostics_plot = FACETS.diagnostics_plot
```
Please just double check this is being produced. For whatever reason, when I ran with Jake's old WDL, these files were spuriously small and could not be rendered.

brendanreardon replied:
Which files in particular?

brendanreardon replied:
When the files were just super small, was it for cases that had NA purity calls? If that is the case, then the PDFs may not have been produced and may have just been preallocated as WDL / Cromwell can't handle optional outputs.

mengxiao replied:
I'm specifically referring to the diagnostic plots, which were 11B unrenderable files in all cases for me, even when there was a purity/ploidy solution. I think if in your test runs you're getting an actual figure, then the problem doesn't apply anymore.

brendanreardon replied:
Oh weird. I've only observed that when purity is NA. Let's look into it if it comes up again.

mengxiao commented on facets.wdl:
```diff
@@ -124,41 +127,51 @@ task Pileup {
-    Int? diskGB
+    Int? preemptible_attempts
+    Int? memoryGB = 4
+    Int? diskGB = ceil(1.1 * (size(cncf, "G"))) + 20
```
Not important, but I hope that this particular python script has no need for the 20 GB constant

brendanreardon commented:
- Added more information and made author information more explicit in the README. 
- Added additional outputs for number of segments and number of segments with NAs.
- Filling NAs with zeros instead of ones when calculating MCN. Also subsetting for autosomal (oops, should have done this before).
- Made pileup vcf not an optional input in the WDL.
- Changed base disk space usage of infer wgd

brendanreardon commented:
Merging pull request. Method is publicly available on FireCloud: https://portal.firecloud.org/#methods/vanallenlab/facets/
